Clean Any interface

robert-zaremba opened the pull request:
## Description

`Any` is an object of handling interface serialization with Protocol Buffers.
We have few methods to create and serialize a concrete value in Any object:
+  `Any.Pack` method
+ `PackAny`  constructor (note: this is different than the line above) 
+ `NewAny*` constructors
+ `UnsafePackAny`
+ `AsAny` (the IntoAny interface)

**GOALS**: 
+ it's good to reduce the public API, especially when it's used for internal purposes. Exposing that API could lead to bad patterns, and will increase the maintenance burden for future SDK releases.
+ try to remove workarounds with `AsAny`

## Summary
+ made `Any.Pack` method private. This method is not needed in a public API, because we have specialized
  functions for createing new Any values. It could be used inappropriately with existing
  Any objects. 
+ removed `PackAny` function.

### References

+ https://github.com/cosmos/cosmos-sdk/issues/7479

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

robert-zaremba commented:
I also removed the `IntoAny` interface - IIRC it was added as a workaround for simulation tests. 
`client` today is using `UnsafePackAny` for handling the interface, so we don't need `IntoAny` at all.

amaurymartiny commented on x/auth/legacy/legacytx/stdtx.go:
```diff
@@ -16,10 +16,9 @@ import (
-	_ sdk.FeeTx          = (*StdTx)(nil)
+	_ sdk.Tx         = (*StdTx)(nil)
+	_ sdk.TxWithMemo = (*StdTx)(nil)
+	_ sdk.FeeTx      = (*StdTx)(nil)
```
shouldn't StdTx also implement ProtoTxProvider?

TxResponse has a tx field of type Any to handle StdTx too, afaiu. It would be good to add a test (or maybe it already exists) to:
- send a StdTx via the REST endpoint
- query the tx by hash
- decode the Any field on TxResponse

robert-zaremba replied:
`ProtoTxProvider` should return a proto transaction - it's used to make sure that we can provide a proto transaction. Initially I was thinking to add that to the `TxBuilder` interface, but I realized that `StdTx` needs to be different. Am I missing something?

amaurymartiny commented on types/tx/types.go:
```diff
@@ -197,3 +197,8 @@ func RegisterInterfaces(registry codectypes.InterfaceRegistry) {
 }
+
+// ProtoTxProvider is a type which can provide a proto transaction.
+type ProtoTxProvider interface {
```
It feels like we're replacing the IntoAny hack with a ProtoTxProvider hack. If you think it's cleaner, then go for it. I don't have a strong preference.

What I would like, for 0.41, is to remove StdTx, so that the Any on TxResponse can be a tx.Tx, and we can remove this hack completely.

robert-zaremba replied:
`IntoAny` was a hack for extracting and serializing a wrapped value. With new constructors we don't need it anymore. `ProtoTxProvider` is not related to `Any`. It's an independent interface - that's why I think it's better. Check for example here: https://github.com/cosmos/cosmos-sdk/pull/8167/files#diff-61a9cb62b0e38da35032d7bc195c38b2c35c342c7302bf6dde5a483f63a96186R458  - `service_test.go:txBuilderToProtoTx`

amaurymartiny reviewed (commented):
OK with the part to make packAny private.

Not strong opinion about ProtoTxProvider, it's not worse or better than the status quo imo.

robert-zaremba commented on codec/types/any.go:
```diff
@@ -58,15 +58,18 @@ type Any struct {
+	if v == nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrPackAny, "Expecting non nil value to create a new Any")
+		// TODO, second option
+		// return nil, nil
```
Me too, especially that the method indicates that we want to create something with a value. Now (in master) it `panic` so, that should be changed anyway.

robert-zaremba commented on x/staking/types/msg.go:
```diff
@@ -31,12 +31,16 @@ var (
+	var pkAny *codectypes.Any
+	if pubKey != nil {
+		var err error
+		if pkAny, err = codectypes.NewAnyWithValue(pubKey); err != nil {
```
You mean removing `PubKeyToAny` ? I was thinking about the same thing. Not sure why it was added.

robert-zaremba replied:
Removed `tx.PubKeyToAny`

amaurymartiny commented on x/auth/client/query.go:
```diff
@@ -134,26 +135,22 @@ func getBlocksForTxResults(clientCtx client.Context, resTxs []*ctypes.ResultTx)
-	tx, err := txConfig.TxDecoder()(txBytes)
+func txToAny(txConfig client.TxConfig, txBytes []byte) (*codectypes.Any, error) {
+	txb, err := txConfig.TxDecoder()(txBytes)
 	if err != nil {
 		return nil, err
 	}
-
-	anyTx, ok := tx.(codectypes.IntoAny)
+	p, ok := txb.(tx.ProtoTxProvider)
```
I think the edge case is when using a TxConfig that's a StdTxConfig, this will fail because StdTx doesn't implement ProtoTxProvider.

iirc, I believe that's why I initially chose the IntoAny hack, because `Any` is something that both wrapper and StdTx understand.

In practice, I think having an amino-only chain doesn't work anyways, so maybe it's fine to drop support for StdTxConfig in this case?

robert-zaremba replied:
Thanks for note. So what we need is a `Tx` (interface) provider. I will have a look into it tomorrow.

amaurymartiny commented on codec/types/any.go:
```diff
@@ -139,8 +125,3 @@ func (any *Any) GetCachedValue() interface{} {
 }
-
-// IntoAny represents a type that can be wrapped into an Any.
-type IntoAny interface {
```
As discussed yesterday, maybe we could rename it, e.g. `TxBuilderToAny`, and move it somewhere less top-level.

robert-zaremba replied:
?, this interface is moved and it's private.

amaurymartiny replied:
I wrote this comment before our conversation last evening.

robert-zaremba commented:
## Summary

+ `Any.Pack` is private 
+ Removed `PackAny` constructor and workarounds with `IntoAny` where the only thing we need is to get the proto Tx (`tx.Tx`)
+ I rolled back `AsAny` removal, but I removed `IntoAny` interface definition from `codectypes`. `AsAny` is only used to create a transaction object from `wrapper` (TxBuilder) or `StdTxConfig`. This is used by the `QueryTx` client API (when we want to return a transaction by tx hash).  NOTE: `StdTxConfig` is officially deprecated. PS: (funny fact): the `wrapper.AsAny` has _GetProtoTx returns the tx as a proto.Message. _ comment in master.
+ I keep the `ProtoTxProvider` interface (created in this PR), because it makes it more clear in places where we intentionally need to get a proto `tx.Tx` object (BuildSimTx and  service_test.go ). The code in master was using a workaround with `Any`: wrapper -> UnsafePackAny -> GetCachedValue, which in my opinion is a hacky way just to extract a `tx.Tx` object from the `TxBuilder`
+ removed `tx.PubKeyToAny` - an alias for `NewAnyWithValue`

**Outcome**
In the public API we only keep the `NewAny*` and `UnsafePackAny` constructors. `AsAny` methods are marked deprecated. The `IntoAny` interface became private.

amaurymartiny commented on x/staking/types/msg.go:
```diff
@@ -31,12 +31,16 @@ var (
-	pkAny, err := codectypes.PackAny(pubKey)
-	if err != nil {
-		return nil, err
+	// TODO: update if we decide to allow nil in NewAnyWithValue
```
Did we finally decide to disallow nil? Does that have repercussions in other places in the SDK?

robert-zaremba replied:
Yes, the only impacted places are where we had `PackAny`, because `NewAny*` were panicking on `nil`.

robert-zaremba replied:
BTW, removing the TODO,

amaurymartiny commented on codec/types/any.go:
```diff
@@ -58,15 +58,16 @@ type Any struct {
-	return any, nil
+	bz, err := proto.Marshal(v)
+	return &Any{
+		TypeUrl:     "/" + proto.MessageName(v),
```
can this call NewAnyWithCustomTypeURL with proto.MessageName(v)?

robert-zaremba replied:
Yes, good idea.

amaurymartiny commented on x/auth/legacy/legacytx/stdtx.go:
```diff
@@ -172,7 +171,9 @@ func (tx StdTx) ValidateBasic() error {
-// AsAny implements IntoAny.AsAny.
+// Deprecated: AsAny implements intoAny. It doesn't work for protobuf serialization,
+// so it can't be saved into protobuf configured storage. We are using it only for API
+// compatibility.
```
👍

amaurymartiny commented on x/auth/tx/builder.go:
```diff
@@ -315,11 +315,14 @@ func (w *wrapper) GetTx() authsigning.Tx {
-	// We're sure here that w.tx is a proto.Message, so this will call
-	// codectypes.NewAnyWithValue under the hood.
-	return codectypes.UnsafePackAny(w.tx)
+func (w *wrapper) GetProtoTx() *tx.Tx {
+	return w.tx
+}
```
I don't think a TxBuilder -> protoTx converter is ever needed.

As I see, we use `ProtoTxProvider` in 2 places:
- in BuildSimTx, for the tx simulate endpoint,
- in a test for said simulate endpoint.

That makes me think that maybe the simulate endpoint is not ideal:
https://github.com/cosmos/cosmos-sdk/blob/f602e9e9776ea37d5b1ac46b724cbf81dff4dd90/proto/cosmos/tx/v1beta1/service.proto#L14-L20

Maybe it should just be `tx_bytes` here, like in broadcast. Thoughts? cc @aaronc

robert-zaremba replied:
I prefer using a object types in proto messages rather than marshaled bytes. The whole point of message is to define structure with object types and let the protocol buffers do the binary / bytes rather than us hacking in bytes of object into the protocol.  This is more bullet / bug proof and forward looking for maintenance.

amaurymartiny replied:
Sounds good too 👍. In this case, could we make `ProtoTxProvider` somewhat local to tx simulations? Since that's the only place that TxBuilder -> protoTx conversion is needed.

robert-zaremba replied:
OK, moved to x/auth/tx

amaurymartiny reviewed (approved):
pre-approving.

The only remaining item is that i conceptually think a txBuilder->protoTx provider is not needed in general, and would like to scope this conversion to tx simulate (only place where it's needed) https://github.com/cosmos/cosmos-sdk/pull/8167#discussion_r545793475.
